[macOS] Refactor top content inset logic to account for content insets on all rect edges

whsieh opened the pull request:
#### fc41c7f03db40c81cb26976470ef16abbe61fff2
<pre>
[macOS] Refactor top content inset logic to account for content insets on all rect edges
<a href="https://bugs.webkit.org/show_bug.cgi?id=287128">https://bugs.webkit.org/show_bug.cgi?id=287128</a>

Reviewed by Simon Fraser and Tim Horton.

Currently, the `-_topContentInset` SPI on macOS is propagated through WebKit into WebCore as a
single `float` value through `Page`, `FrameView`, various scrolling/compositing-related classes,
and many other codepaths. In preparation for exposing API to control content insets on all sides of
a webpage on macOS, this PR refactors support for top content inset in WebKit and WebCore to plumb
a full `RectEdges&lt;float&gt;` (i.e. `FloatBoxExtent`) through the engine, instead of just a single
`float` for the top content inset.

Note that there should be no change in behavior yet, since the ability for clients to set non-zero
left/right/bottom content insets does not exist yet. In a subsequent patch, I&apos;ll add this capability
along with tests to exercise it.

See below for more details.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Drive-by fix: add `FrameView.cpp` to the Xcode project. This source file is currently compiled as
part of unified sources, but is absent from Xcode (making text searches skip this file).

* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::drawBounds):
(WebCore::InspectorOverlay::drawRulers):
(WebCore::InspectorOverlay::drawElementTitle):
(WebCore::InspectorOverlay::buildGridOverlay):

Swap `topContentInset` for `obscuredContentInsets`.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::obscuredContentInsets const):
(WebCore::FrameView::topContentInset const): Deleted.

Replace `topContentInset` with `obscuredContentInsets`; rename `TopContentInsetType` to `InsetType`.

* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::obscuredContentInsetsDidChange):

Rename `topContentInsetDidChange` → `obscuredContentInsetsDidChange`.

(WebCore::LocalFrameView::yPositionForInsetClipLayer):
(WebCore::LocalFrameView::yPositionForHeaderLayer):
(WebCore::LocalFrameView::yPositionForFooterLayer):
(WebCore::LocalFrameView::positionForRootContentLayer):
(WebCore::LocalFrameView::positionForRootContentLayer const):

Pass in `obscuredContentInsets().top()` in place of `topContentInset()`. Note that we might need
more refactoring in this area in a subsequent patch to make it properly handle left/right insets.

(WebCore::LocalFrameView::topContentInsetDidChange): Deleted.
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setObscuredContentInsets):
(WebCore::Page::setTopContentInset): Deleted.

Rename `setTopContentInset` → `setObscuredContentInsets`, and make it take a `FloatBoxExtent`. Note
that the `m_obscuredInsets` member already exists on `Page`, but is currently iOS-only, and reflects
the value of `_obscuredInset` on `WKWebView`. This is actually slightly distinct from this
refactored notion of `obscuredContentInsets` on macOS, which represents extra margins that surround
the `Page`, are intended to be obscured by the WebKit client&apos;s UI at scroll extents, and also
requires WebCore to move its layers to accomodate the inset content.

* Source/WebCore/page/Page.h:
(WebCore::Page::obscuredInsets const):
(WebCore::Page::setObscuredInsets):
(WebCore::Page::obscuredContentInsets const):
(WebCore::Page::topContentInset const): Deleted.
* Source/WebCore/page/ResourceUsageOverlay.cpp:
(WebCore::ResourceUsageOverlay::mouseEvent):
* Source/WebCore/page/mac/ServicesOverlayController.mm:
(WebCore::ServicesOverlayController::buildPhoneNumberHighlights):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::reconcileScrollingState):
(WebCore::AsyncScrollingCoordinator::setFrameScrollingNodeState):

Plumb content insets through `ScrollingStateFrameScrollingNode` and related classes.

* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::ScrollingStateFrameScrollingNode):
(WebCore::ScrollingStateFrameScrollingNode::applicableProperties const):
(WebCore::ScrollingStateFrameScrollingNode::setObscuredContentInsets):
(WebCore::ScrollingStateFrameScrollingNode::dumpProperties const):
(WebCore::ScrollingStateFrameScrollingNode::setTopContentInset): Deleted.
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.h:

Rename the `TopContentInset` property flag → `ObscuredContentInsets`.

* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::mainFrameObscuredContentInsets const):
(WebCore::ScrollingTree::mainFrameTopContentInset const): Deleted.
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeFrameScrollingNode.cpp:
(WebCore::ScrollingTreeFrameScrollingNode::commitStateBeforeChildren):
(WebCore::ScrollingTreeFrameScrollingNode::viewToContentsOffset const):
(WebCore::ScrollingTreeFrameScrollingNode::dumpProperties const):

Augment this debugging description to show non-zero top/bottom/left/right insets separately.

* Source/WebCore/page/scrolling/ScrollingTreeFrameScrollingNode.h:
* Source/WebCore/page/scrolling/coordinated/ScrollingTreeFrameScrollingNodeCoordinated.cpp:
(WebCore::ScrollingTreeFrameScrollingNodeCoordinated::repositionRelatedLayers):
* Source/WebCore/page/scrolling/mac/ScrollingTreeFrameScrollingNodeMac.mm:
(WebCore::ScrollingTreeFrameScrollingNodeMac::repositionRelatedLayers):
* Source/WebCore/platform/BoxExtents.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKThumbnailViewInternal.h.

Add a new `WebCore/platform` header that defines several box extent types used throughout WebCore
and WebKit:

- FloatBoxExtent
- IntBoxExtent
- LayoutBoxExtent

* Source/WebCore/platform/LengthBox.h:
* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::sizeForUnobscuredContent const):

Account for the left inset.

(WebCore::ScrollView::frameRectShrunkByInset const):

Account for insets on all sides.

(WebCore::ScrollView::documentScrollPositionRelativeToViewOrigin const):
(WebCore::ScrollView::updateScrollbars):
(WebCore::ScrollView::rootViewToTotalContents const):
(WebCore::ScrollView::scrollCornerRect const):
(WebCore::ScrollView::platformContentInsets const):
(WebCore::ScrollView::platformSetContentInsets):
(WebCore::ScrollView::platformTopContentInset const): Deleted.
(WebCore::ScrollView::platformSetTopContentInset): Deleted.
* Source/WebCore/platform/ScrollView.h:
(WebCore::ScrollView::obscuredContentInsets const):
(WebCore::ScrollView::topContentInset const): Deleted.
* Source/WebCore/platform/graphics/TiledBacking.h:
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::setObscuredContentInsets):
(WebCore::TileController::setTopContentInset): Deleted.

Replace more `topContentInset` with `obscuredContentInsets`.

* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebCore/platform/graphics/ca/TileCoverageMap.cpp:
(WebCore::TileCoverageMap::TileCoverageMap):
(WebCore::TileCoverageMap::update):
* Source/WebCore/platform/ios/ScrollViewIOS.mm:
(WebCore::ScrollView::platformContentInsets const):
(WebCore::ScrollView::platformSetContentInsets):
(WebCore::ScrollView::platformTopContentInset const): Deleted.
(WebCore::ScrollView::platformSetTopContentInset): Deleted.
* Source/WebCore/platform/mac/ScrollViewMac.mm:
(WebCore::ScrollView::platformContentInsets const):
(WebCore::ScrollView::platformSetContentInsets):
(WebCore::ScrollView::platformTopContentInset const): Deleted.
(WebCore::ScrollView::platformSetTopContentInset): Deleted.
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::calculateBackgroundImageGeometry):

Account for the left inset.

* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateBacking):
(WebCore::RenderLayerCompositor::positionForClipLayer const):
(WebCore::RenderLayerCompositor::updateLayerForBottomOverhangArea):
(WebCore::RenderLayerCompositor::updateLayerForHeader):
(WebCore::RenderLayerCompositor::updateLayerForFooter):
(WebCore::RenderLayerCompositor::updateSizeAndPositionForOverhangAreaLayer):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(WebKit::dump):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _recalculateViewportSizesWithMinimumViewportInset:maximumViewportInset:throwOnInvalidInput:]):

Account for the left content inset.

* Source/WebKit/UIProcess/API/Cocoa/_WKThumbnailView.mm:
(-[_WKThumbnailView requestSnapshot]):
(-[_WKThumbnailView _viewWasUnparented]):
(-[_WKThumbnailView _viewWasParented]):
(-[_WKThumbnailView setScale:]):
(-[_WKThumbnailView _setSublayerTranslation:]):
(-[_WKThumbnailView _setSublayerVerticalTranslationAmount:]): Deleted.

Rename `sublayerVerticalTranslationAmount` to just `sublayerTranslation`, which is now a `CGPoint`
that represents both `x` and `y` offsets.

* Source/WebKit/UIProcess/API/Cocoa/_WKThumbnailViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _setTopContentInset:]):
(-[WKWebView _topContentInset]):
(-[WKWebView _setTopContentInset:immediate:]):

Keep these APIs as-is, but implement them in terms of `FloatBoxExtent`-backed content insets on the
`WebViewImpl`.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::performMouseInteraction):

Account for the left inset.

* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::viewportLocationToWindowLocation):
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::obscuredContentInsetsDidChange):
(WebKit::PageClientImplCocoa::topContentInsetDidChange): Deleted.

Rename `topContentInsetDidChange` to `obscuredContentInsetsDidChange`.

* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::obscuredContentInsetsDidChange):
(WebKit::PageClient::computeAutomaticTopObscuredInset):
(WebKit::PageClient::topContentInsetDidChange): Deleted.
(WebKit::PageClient::computeAutomaticTopContentInset): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::dispatchSetObscuredContentInsets):
(WebKit::RemoteLayerTreeDrawingAreaProxy::dispatchSetTopContentInset): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::sendUpdateGeometry):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::obscuredContentInsets const):
(WebKit::RemoteScrollingCoordinatorProxy::topContentInset const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::layoutBannerLayers):
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::dispatchSetObscuredContentInsets):
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::dispatchSetTopContentInset): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingNodeForPoint):
* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::shouldUseSnapshotForSize):

Make this account for non-zero inset as well.

* Source/WebKit/UIProcess/ViewGestureController.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setObscuredContentInsets):
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::setTopContentInset): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(WebKit::WKWebViewState::applyTo):
(WebKit::WKWebViewState::store):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::computeAutomaticTopObscuredInset):
(WebKit::PageClientImpl::computeAutomaticTopContentInset): Deleted.
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::handleMagnificationGestureEvent):
(WebKit::ViewGestureController::beginSwipeGesture):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.h:
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController enterFullScreen:completionHandler:]):
(-[WKFullScreenWindowController finishedEnterFullScreenAnimation:]):
(-[WKFullScreenWindowController finishedExitFullScreenAnimationAndExitImmediately:]):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::updateContentInsetsIfAutomatic):
(WebKit::WebPageProxy::setObscuredContentInsetsAsync):
(WebKit::WebPageProxy::pendingOrActualObscuredContentInsets const):
(WebKit::WebPageProxy::scheduleSetObscuredContentInsetsDispatch):
(WebKit::WebPageProxy::dispatchSetObscuredContentInsets):
(WebKit::WebPageProxy::setTopContentInsetAsync): Deleted.
(WebKit::WebPageProxy::pendingOrActualTopContentInset const): Deleted.
(WebKit::WebPageProxy::scheduleSetTopContentInsetDispatch): Deleted.
(WebKit::WebPageProxy::dispatchSetTopContentInset): Deleted.
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::obscuredContentInsets const):
(WebKit::WebViewImpl::setObscuredContentInsets):
(WebKit::WebViewImpl::flushPendingObscuredContentInsetChanges):
(WebKit::WebViewImpl::endDeferringViewInWindowChanges):
(WebKit::WebViewImpl::endDeferringViewInWindowChangesSync):
(WebKit::WebViewImpl::prepareForMoveToWindow):
(WebKit::WebViewImpl::takeViewSnapshot):

Make this account for insets on all sides.

(WebKit::WebViewImpl::topContentInset const): Deleted.
(WebKit::WebViewImpl::setTopContentInset): Deleted.
(WebKit::WebViewImpl::flushPendingTopContentInset): Deleted.

Rename `flushPendingTopContentInset` to `flushPendingObscuredContentInsetChanges`.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:

Remove an unused member, `m_topContentInset`.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setPageScaleFactor):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::obscuredContentInsetsDidChange):
(WebKit::PluginView::topContentInsetDidChange): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.cpp:
(WebKit::DrawingArea::scaleViewToFitDocumentIfNeeded):

Add a couple of FIXMEs to (possibly) account for left content inset in fixed layout mode, in a
future patch.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
(WebKit::WebPage::setObscuredContentInsetsFenced):
(WebKit::WebPage::obscuredContentInsets const):
(WebKit::WebPage::setObscuredContentInsets):
(WebKit::WebPage::setTopContentInsetFenced): Deleted.
(WebKit::WebPage::setTopContentInset): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPageTesting.cpp:
(WebKit::WebPageTesting::setTopContentInset):

Make this testing-only helper set only the top obscured content inset on the page.

* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _setTopContentInsetForTesting:]):

Canonical link: <a href="https://commits.webkit.org/290068@main">https://commits.webkit.org/290068@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c745b22fe7865421ca00e55b357616d8cbaa5314

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88651 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8170 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43098 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93613 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39405 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90702 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8557 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16354 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68355 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26052 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91653 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6554 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80182 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48724 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6314 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34592 "Found 3 new test failures: fonts/monospace.html fonts/sans-serif.html fonts/serif.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38513 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76684 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35492 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95451 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15826 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11588 "Found 1 new test failure: imported/w3c/web-platform-tests/WebCryptoAPI/import_export/symmetric_importKey.https.any.worker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77223 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16082 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76041 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76506 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18894 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20886 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19308 "Found 5 new test failures: http/tests/webarchive/cross-origin-stylesheet-crash.html http/tests/webarchive/test-css-url-encoding-shift-jis.html http/tests/webarchive/test-css-url-encoding-utf-8.html http/tests/webarchive/test-css-url-encoding.html http/tests/webarchive/test-preload-resources.html (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8865 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15842 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21150 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15583 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19032 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17365 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/WebPage/DrawingArea.cpp:
```diff
@@ -228,7 +228,8 @@ void DrawingArea::scaleViewToFitDocumentIfNeeded()
                 documentWidth = std::ceil(viewWidth / viewScale);
             }
-            IntSize fixedLayoutSize(documentWidth, std::ceil((webPage->size().height() - webPage->corePage()->topContentInset()) / viewScale));
+            // FIXME: Account for left content insets.
```
These may be load-bearing FIXMEs in tiled fullscreen

whsieh replied:
Oh, definitely — I'll be working through these as I continue fleshing out support for the new API (next on the list are the FIXMEs in `positionFor{InsetClipLayer|ContentLayer}`).

Thanks for the reviews!

whsieh committed:
[macOS] Refactor top content inset logic to account for content insets on all rect edges
https://bugs.webkit.org/show_bug.cgi?id=287128

Reviewed by Simon Fraser and Tim Horton.

Currently, the `-_topContentInset` SPI on macOS is propagated through WebKit into WebCore as a
single `float` value through `Page`, `FrameView`, various scrolling/compositing-related classes,
and many other codepaths. In preparation for exposing API to control content insets on all sides of
a webpage on macOS, this PR refactors support for top content inset in WebKit and WebCore to plumb
a full `RectEdges<float>` (i.e. `FloatBoxExtent`) through the engine, instead of just a single
`float` for the top content inset.

Note that there should be no change in behavior yet, since the ability for clients to set non-zero
left/right/bottom content insets does not exist yet. In a subsequent patch, I'll add this capability
along with tests to exercise it.

See below for more details.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Drive-by fix: add `FrameView.cpp` to the Xcode project. This source file is currently compiled as
part of unified sources, but is absent from Xcode (making text searches skip this file).

* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::drawBounds):
(WebCore::InspectorOverlay::drawRulers):
(WebCore::InspectorOverlay::drawElementTitle):
(WebCore::InspectorOverlay::buildGridOverlay):

Swap `topContentInset` for `obscuredContentInsets`.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::obscuredContentInsets const):
(WebCore::FrameView::topContentInset const): Deleted.

Replace `topContentInset` with `obscuredContentInsets`; rename `TopContentInsetType` to `InsetType`.

* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::obscuredContentInsetsDidChange):

Rename `topContentInsetDidChange` → `obscuredContentInsetsDidChange`.

(WebCore::LocalFrameView::yPositionForInsetClipLayer):
(WebCore::LocalFrameView::yPositionForHeaderLayer):
(WebCore::LocalFrameView::yPositionForFooterLayer):
(WebCore::LocalFrameView::positionForRootContentLayer):
(WebCore::LocalFrameView::positionForRootContentLayer const):

Pass in `obscuredContentInsets().top()` in place of `topContentInset()`. Note that we might need
more refactoring in this area in a subsequent patch to make it properly handle left/right insets.

(WebCore::LocalFrameView::topContentInsetDidChange): Deleted.
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setObscuredContentInsets):
(WebCore::Page::setTopContentInset): Deleted.

Rename `setTopContentInset` → `setObscuredContentInsets`, and make it take a `FloatBoxExtent`. Note
that the `m_obscuredInsets` member already exists on `Page`, but is currently iOS-only, and reflects
the value of `_obscuredInset` on `WKWebView`. This is actually slightly distinct from this
refactored notion of `obscuredContentInsets` on macOS, which represents extra margins that surround
the `Page`, are intended to be obscured by the WebKit client's UI at scroll extents, and also
requires WebCore to move its layers to accomodate the inset content.

* Source/WebCore/page/Page.h:
(WebCore::Page::obscuredInsets const):
(WebCore::Page::setObscuredInsets):
(WebCore::Page::obscuredContentInsets const):
(WebCore::Page::topContentInset const): Deleted.
* Source/WebCore/page/ResourceUsageOverlay.cpp:
(WebCore::ResourceUsageOverlay::mouseEvent):
* Source/WebCore/page/mac/ServicesOverlayController.mm:
(WebCore::ServicesOverlayController::buildPhoneNumberHighlights):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::reconcileScrollingState):
(WebCore::AsyncScrollingCoordinator::setFrameScrollingNodeState):

Plumb content insets through `ScrollingStateFrameScrollingNode` and related classes.

* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::ScrollingStateFrameScrollingNode):
(WebCore::ScrollingStateFrameScrollingNode::applicableProperties const):
(WebCore::ScrollingStateFrameScrollingNode::setObscuredContentInsets):
(WebCore::ScrollingStateFrameScrollingNode::dumpProperties const):
(WebCore::ScrollingStateFrameScrollingNode::setTopContentInset): Deleted.
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.h:

Rename the `TopContentInset` property flag → `ObscuredContentInsets`.

* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::mainFrameObscuredContentInsets const):
(WebCore::ScrollingTree::mainFrameTopContentInset const): Deleted.
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeFrameScrollingNode.cpp:
(WebCore::ScrollingTreeFrameScrollingNode::commitStateBeforeChildren):
(WebCore::ScrollingTreeFrameScrollingNode::viewToContentsOffset const):
(WebCore::ScrollingTreeFrameScrollingNode::dumpProperties const):

Augment this debugging description to show non-zero top/bottom/left/right insets separately.

* Source/WebCore/page/scrolling/ScrollingTreeFrameScrollingNode.h:
* Source/WebCore/page/scrolling/coordinated/ScrollingTreeFrameScrollingNodeCoordinated.cpp:
(WebCore::ScrollingTreeFrameScrollingNodeCoordinated::repositionRelatedLayers):
* Source/WebCore/page/scrolling/mac/ScrollingTreeFrameScrollingNodeMac.mm:
(WebCore::ScrollingTreeFrameScrollingNodeMac::repositionRelatedLayers):
* Source/WebCore/platform/BoxExtents.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKThumbnailViewInternal.h.

Add a new `WebCore/platform` header that defines several box extent types used throughout WebCore
and WebKit:

- FloatBoxExtent
- IntBoxExtent
- LayoutBoxExtent

* Source/WebCore/platform/LengthBox.h:
* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::sizeForUnobscuredContent const):

Account for the left inset.

(WebCore::ScrollView::frameRectShrunkByInset const):

Account for insets on all sides.

(WebCore::ScrollView::documentScrollPositionRelativeToViewOrigin const):
(WebCore::ScrollView::updateScrollbars):
(WebCore::ScrollView::rootViewToTotalContents const):
(WebCore::ScrollView::scrollCornerRect const):
(WebCore::ScrollView::platformContentInsets const):
(WebCore::ScrollView::platformSetContentInsets):
(WebCore::ScrollView::platformTopContentInset const): Deleted.
(WebCore::ScrollView::platformSetTopContentInset): Deleted.
* Source/WebCore/platform/ScrollView.h:
(WebCore::ScrollView::obscuredContentInsets const):
(WebCore::ScrollView::topContentInset const): Deleted.
* Source/WebCore/platform/graphics/TiledBacking.h:
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::setObscuredContentInsets):
(WebCore::TileController::setTopContentInset): Deleted.

Replace more `topContentInset` with `obscuredContentInsets`.

* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebCore/platform/graphics/ca/TileCoverageMap.cpp:
(WebCore::TileCoverageMap::TileCoverageMap):
(WebCore::TileCoverageMap::update):
* Source/WebCore/platform/ios/ScrollViewIOS.mm:
(WebCore::ScrollView::platformContentInsets const):
(WebCore::ScrollView::platformSetContentInsets):
(WebCore::ScrollView::platformTopContentInset const): Deleted.
(WebCore::ScrollView::platformSetTopContentInset): Deleted.
* Source/WebCore/platform/mac/ScrollViewMac.mm:
(WebCore::ScrollView::platformContentInsets const):
(WebCore::ScrollView::platformSetContentInsets):
(WebCore::ScrollView::platformTopContentInset const): Deleted.
(WebCore::ScrollView::platformSetTopContentInset): Deleted.
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::calculateBackgroundImageGeometry):

Account for the left inset.

* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateBacking):
(WebCore::RenderLayerCompositor::positionForClipLayer const):
(WebCore::RenderLayerCompositor::updateLayerForBottomOverhangArea):
(WebCore::RenderLayerCompositor::updateLayerForHeader):
(WebCore::RenderLayerCompositor::updateLayerForFooter):
(WebCore::RenderLayerCompositor::updateSizeAndPositionForOverhangAreaLayer):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(WebKit::dump):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _recalculateViewportSizesWithMinimumViewportInset:maximumViewportInset:throwOnInvalidInput:]):

Account for the left content inset.

* Source/WebKit/UIProcess/API/Cocoa/_WKThumbnailView.mm:
(-[_WKThumbnailView requestSnapshot]):
(-[_WKThumbnailView _viewWasUnparented]):
(-[_WKThumbnailView _viewWasParented]):
(-[_WKThumbnailView setScale:]):
(-[_WKThumbnailView _setSublayerTranslation:]):
(-[_WKThumbnailView _setSublayerVerticalTranslationAmount:]): Deleted.

Rename `sublayerVerticalTranslationAmount` to just `sublayerTranslation`, which is now a `CGPoint`
that represents both `x` and `y` offsets.

* Source/WebKit/UIProcess/API/Cocoa/_WKThumbnailViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _setTopContentInset:]):
(-[WKWebView _topContentInset]):
(-[WKWebView _setTopContentInset:immediate:]):

Keep these APIs as-is, but implement them in terms of `FloatBoxExtent`-backed content insets on the
`WebViewImpl`.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::performMouseInteraction):

Account for the left inset.

* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::viewportLocationToWindowLocation):
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::obscuredContentInsetsDidChange):
(WebKit::PageClientImplCocoa::topContentInsetDidChange): Deleted.

Rename `topContentInsetDidChange` to `obscuredContentInsetsDidChange`.

* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::obscuredContentInsetsDidChange):
(WebKit::PageClient::computeAutomaticTopObscuredInset):
(WebKit::PageClient::topContentInsetDidChange): Deleted.
(WebKit::PageClient::computeAutomaticTopContentInset): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::dispatchSetObscuredContentInsets):
(WebKit::RemoteLayerTreeDrawingAreaProxy::dispatchSetTopContentInset): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::sendUpdateGeometry):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::obscuredContentInsets const):
(WebKit::RemoteScrollingCoordinatorProxy::topContentInset const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::layoutBannerLayers):
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::dispatchSetObscuredContentInsets):
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::dispatchSetTopContentInset): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingNodeForPoint):
* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::shouldUseSnapshotForSize):

Make this account for non-zero inset as well.

* Source/WebKit/UIProcess/ViewGestureController.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setObscuredContentInsets):
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::setTopContentInset): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(WebKit::WKWebViewState::applyTo):
(WebKit::WKWebViewState::store):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::computeAutomaticTopObscuredInset):
(WebKit::PageClientImpl::computeAutomaticTopContentInset): Deleted.
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::handleMagnificationGestureEvent):
(WebKit::ViewGestureController::beginSwipeGesture):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.h:
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController enterFullScreen:completionHandler:]):
(-[WKFullScreenWindowController finishedEnterFullScreenAnimation:]):
(-[WKFullScreenWindowController finishedExitFullScreenAnimationAndExitImmediately:]):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::updateContentInsetsIfAutomatic):
(WebKit::WebPageProxy::setObscuredContentInsetsAsync):
(WebKit::WebPageProxy::pendingOrActualObscuredContentInsets const):
(WebKit::WebPageProxy::scheduleSetObscuredContentInsetsDispatch):
(WebKit::WebPageProxy::dispatchSetObscuredContentInsets):
(WebKit::WebPageProxy::setTopContentInsetAsync): Deleted.
(WebKit::WebPageProxy::pendingOrActualTopContentInset const): Deleted.
(WebKit::WebPageProxy::scheduleSetTopContentInsetDispatch): Deleted.
(WebKit::WebPageProxy::dispatchSetTopContentInset): Deleted.
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::obscuredContentInsets const):
(WebKit::WebViewImpl::setObscuredContentInsets):
(WebKit::WebViewImpl::flushPendingObscuredContentInsetChanges):
(WebKit::WebViewImpl::endDeferringViewInWindowChanges):
(WebKit::WebViewImpl::endDeferringViewInWindowChangesSync):
(WebKit::WebViewImpl::prepareForMoveToWindow):
(WebKit::WebViewImpl::takeViewSnapshot):

Make this account for insets on all sides.

(WebKit::WebViewImpl::topContentInset const): Deleted.
(WebKit::WebViewImpl::setTopContentInset): Deleted.
(WebKit::WebViewImpl::flushPendingTopContentInset): Deleted.

Rename `flushPendingTopContentInset` to `flushPendingObscuredContentInsetChanges`.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:

Remove an unused member, `m_topContentInset`.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setPageScaleFactor):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::obscuredContentInsetsDidChange):
(WebKit::PluginView::topContentInsetDidChange): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.cpp:
(WebKit::DrawingArea::scaleViewToFitDocumentIfNeeded):

Add a couple of FIXMEs to (possibly) account for left content inset in fixed layout mode, in a
future patch.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
(WebKit::WebPage::setObscuredContentInsetsFenced):
(WebKit::WebPage::obscuredContentInsets const):
(WebKit::WebPage::setObscuredContentInsets):
(WebKit::WebPage::setTopContentInsetFenced): Deleted.
(WebKit::WebPage::setTopContentInset): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPageTesting.cpp:
(WebKit::WebPageTesting::setTopContentInset):

Make this testing-only helper set only the top obscured content inset on the page.

* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _setTopContentInsetForTesting:]):

Canonical link: https://commits.webkit.org/290068@main
